dali/operators/random/rng_base.h

```diff
@@ -53,18 +55,37 @@ class RNGBase : public Operator<Backend> {
       rng_ = rng;
     } else {
       static_assert(std::is_same_v<Backend, GPUBackend>);
-      DALI_FAIL("Checkpointing is not implemented for GPU random operators. ");
+      const auto &states_cpu = cpt.CheckpointState<std::shared_ptr<curandState>>();
+      backend_data_.randomizer_.set_states(states_cpu.get());
     }
   }
 
   std::string SerializeCheckpoint(const OpCheckpoint &cpt) const override {
-    const auto &state = cpt.CheckpointState<BatchRNG<std::mt19937_64>>();
-    return SnapshotSerializer().Serialize(state.ToVector());
+    if constexpr (std::is_same_v<Backend, CPUBackend>) {
+      const auto &state = cpt.CheckpointState<BatchRNG<std::mt19937_64>>();
+      return SnapshotSerializer().Serialize(state.ToVector());
+    } else {
+      static_assert(std::is_same_v<Backend, GPUBackend>);
+      const auto &ptr_gpu = cpt.CheckpointState<std::shared_ptr<curandState>>();
+      size_t n = backend_data_.randomizer_.length();
```

Comment:
Can the `n` change at any point? I saw some checks for batch size in CPU code, should the `n` be the part of the checkpoint itself?

Reply:
I suppose that the only way this can happen is when the user tries to restore a pipeline from a checkpoint created with a pipeline with different batch size. 

The issue is not only about batch size, but about all settings of the pipeline and operators. Checking it here doesn't make much sense to me. I'll think about a way to check such things strictly somewhere at the higher level (maybe we should have something some kind of "pipeline hash" and check if it matches?). 

Thank you for reminding me of that - while it's kind of ok to have such behavior in the "experimental" checkpointing, the final implementation should protect the users from harming themselves

Reply:
After offline discussion with @stiepan we decided that there aren't that many ways of the user harming themselves, so just adding an assert here is quite a good idea. I'll add one in a moment

Reply:
Adding that assert required surprisingly many changes (see https://github.com/NVIDIA/DALI/pull/5148/commits/af4ccdca32fd762d1b9629138d97b5ac39616d11), because I was not keeping the number of states in the checkpoint: it was just a pointer to the states and the length was deduced from current randomizer's length.

I changed the checkpoint type from `shared_ptr<curandState>` to `curand_states`, which is exactly what I wanted: a shared ptr plus length. This allows me to keep track of the length cleanly

Reply:
This change solves my reservation from the first comment, where the state/checkpoint was a pointer to memory, but the length of the memory was tied to the actual state of the operator.